### PR TITLE
 Added missing s3_settings attributes

### LINKED ISSUE #8000: DMS endpoint s3 target to support extra_connection_attributes
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

DMS has introduced new attributes in s3_settings and api doc will be updated end of this week, it can be passed as either s3_settings or extra_connection_attributes. 

Terraform for s3 target doesn't support extra_connection_attributes, it would be nice to add it so new features on endpoint can be utilized through extra_connection_attributes.

https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_dms_endpoint.go#L283

API which supports new attributes

aws dms create-endpoint --endpoint-identifier s3-target-endpoint --engine-name s3 --endpoint-type target
--s3-settings ‘{“ServiceAccessRoleArn”: “your-service-access-ARN”, “DataFormat”: “parquet”, "parquetVersion":"PARQUET_2_0"}’

aws dms create-endpoint   --extra-connection-attributes "ServiceAccessRoleArn:your-service-access-ARN;bucketName=dms-test-321;bucketFolder=s3test;DataFormat=parquet;parquetVersion=PARQUET_2_0;"

### New or Affected Resource(s)

* aws_dms_endpoint

### Potential Terraform Configuration

```hcl
resource "aws_dms_endpoint" "media_s3_endpoint" {
  endpoint_id  = ""
  endpoint_type = "target"
  extra_connection_attributes = "DataFormat=parquet;parquetVersion=PARQUET_2_0;"
  engine_name = "s3"
  s3_settings = {
    service_access_role_arn = ""
    bucket_name= ""
    bucket_folder = ""
    compression_type= "GZIP"
  }

  tags = ""

}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by lyle-nel**
Extend DMS S3 settings to support updated API changes in aws-sdk-go. This is so that we can avoid using extra_connection_attributes.

Full list of supported attributes for s3_settings were found here: [models/apis/dms/2016-01-01/api-2.json](https://github.com/aws/aws-sdk-go/blob/0206d02e95bdde17c7ebdaa3844624ea77ffc5cb/models/apis/dms/2016-01-01/api-2.json#L2055)

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #8000 #8009

### Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

ENHANCEMENTS:
 * resource/aws_dms_endpoint: Added missing `s3_settings` attributes `ServiceAccessRoleArn, ExternalTableDefinition, CsvRowDelimiter, CsvDelimiter, BucketFolder, BucketName, CompressionType, TimestampColumnName, DataFormat, ParquetVersion, EncryptionMode, ServerSideEncryptionKmsKeyId`

### Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAwsDmsEndpoint_S3'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsDmsEndpoint_S3 -timeout 120m
=== RUN   TestAccAwsDmsEndpoint_S3_Csv
=== PAUSE TestAccAwsDmsEndpoint_S3_Csv
=== RUN   TestAccAwsDmsEndpoint_S3_Parquet
=== PAUSE TestAccAwsDmsEndpoint_S3_Parquet
=== CONT  TestAccAwsDmsEndpoint_S3_Csv
=== CONT  TestAccAwsDmsEndpoint_S3_Parquet
--- PASS: TestAccAwsDmsEndpoint_S3_Csv (291.42s)
--- PASS: TestAccAwsDmsEndpoint_S3_Parquet (365.29s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	365.311s
...
```

### CODE REVIEW

**Comment by lyle-nel** [comment]
Hi Folks, is there something that is left to do or can we merge this?

Note that even though the size estimation label says "size/L", it is actually a trivial change. The LOC is high due to the number of additional attributes that were added.

**Comment by gerovermaas** [comment]
Hi @lyle-nel, 

I was happy to see this PR since we were also facing this issue. 

I tested it, but do run into a problem. I have defined the end point as:

```resource "aws_dms_endpoint" "s3_target" {
  endpoint_id   = "s3-target-endpoint"
  endpoint_type = "target"
  engine_name   = "s3"

  extra_connection_attributes = "includeOpForFullLoad=true"

  s3_settings {
    service_access_role_arn = aws_iam_role.s3_target.arn
    bucket_folder           = aws_db_instance.mysql.name
    bucket_name             = aws_s3_bucket.target.id
    # data_format             = "parquet"
    # timestamp_column_name   = "updateTS"

  }
}```

This provisions fine with the 2.44.0 version of the AWS provider, but when I provision it with your version of the AWS provider, testing of the endpoint by AWS fails with the message: "Error Details: [message=Exception while converting from replicate model object Test endpoint, errType=, status=0, errMessage=, errDetails=]"

Enabling the the commented lines (`data_format`, `timestamp_column_name`) does not change it, fails with same message. (I disabled this, since I wanted to verify it still worked without the extract attributes added by this PR)

Using Terraform 0.12.19

Am I missing something?

Regards,
Gero

**Comment by lyle-nel** [comment]
Hi @gerovermaas, I will look into this tomorrow morning (UTC+2) and get back to you as soon as possible.

**Comment by lyle-nel** [comment]
@gerovermaas Thanks I am busy looking into why these additional attributes are showing up. Even in the 2.44.0 version in your example the EnableStatistics attribute is showing up even though that should only happen in the case of writing to parquet.

I have tracked down one potential avenue where the defaults are set to empty strings, but taking that away does not have an effect on the outcome.

I will keep you updated if I learn anything useful. For now, I just need to exclude the possibility that it is anything terraform related, then I can move on to seeing if the issue lies with the aws-go-sdk

**Comment by gerovermaas** [comment]
@lyle-nel Any progress or update on this PR?

**Comment by lyle-nel** [comment]
@gerovermaas I am actively working on this. I don't have much of an update besides that I have written some integration tests in the aws provider that exposes the issue. I am currently writing some tests for the schema helpers in terraform core. It is very unlikely that the issue is there, but I just need to exclude that possibility.

For the sake of transparency, I will summarise the symptom:

Consider the following resource schema:
``` go
Schema: map[string]*schema.Schema{
	"s3_settings": {
		Type:     schema.TypeList,
		Optional: true,
		MaxItems: 1,
		DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
			if old == "1" && new == "0" {
				return true
			}
			return false
		},
		Elem: &schema.Resource{
			Schema: map[string]*schema.Schema{

				"compression_type": {
					Type:     schema.TypeString,
					Optional: true,
					Default:  "NONE",
				},
				"timestamp_column_name": {
					Type:        schema.TypeString,
					Optional:    true,
				}
			},
		},
	},
},
```
It is clear from this that `compression_type` is optional with a default value of `"None"` so we should expect to find that default value in the endpoint configuration. It is also clear that `timestamp_column_name` has no default defined, so it should be omitted entirely from the endpoint configuration. However, what we find is that `timestamp_column_name` is not omitted as expected.

**Comment by lyle-nel** [comment]
@gerovermaas I managed to track down the issue. My tests led me astray, so the solution is pretty straightforward. The reason I know it works without resorting to the integration test to tell me that it works is that I simply inspect the response after doing the deployment.
The problem I am faced with now is how I would write an integration test to test that an attribute is **not** set? I haven't yet found a case where this sort of thing is tested in the aws provider. Which leads me to believe this is either non-trivial, or I am missing something obvious. At any rate, below is the test utility that should work, but doesn't.
```go
func UtilTestCheckResourceAttrNotSet(name, key string) TestCheckFunc {
	return func(s *terraform.State) error {
		is, err := primaryInstanceState(s, name)
		if err != nil {
			return err
		}

		return testCheckResourceAttrNotSet(is, name, key)
	}
}
func testCheckResourceAttrNotSet(is *terraform.InstanceState, name string, key string) error {
	if _, ok := is.Attributes[key]; ok {
		return fmt.Errorf("%s: Attribute '%s' expected not to be set", name, key)
	}

	return nil
}
```
I will push the changes tomorrow morning.

**Comment by gerovermaas** [comment]
Great to read @lyle-nel !

File: aws/resource_aws_dms_endpoint_test.go [hunk 1]
```diff
@@ -95,7 +100,58 @@ func TestAccAwsDmsEndpoint_S3(t *testing.T) {
...
+			{
+				Config: dmsEndpointS3Config(randId),
+				Check: resource.ComposeTestCheckFunc(
+					checkDmsEndpointExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "s3_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "s3_settings.0.bucket_folder", ""),
+					resource.TestCheckResourceAttr(resourceName, "s3_settings.0.bucket_name", "bucket_name"),
+					resource.TestCheckResourceAttr(resourceName, "s3_settings.0.compression_type", "NONE"),
+					resource.TestCheckNoResourceAttr(resourceName, "s3_settings.0.enable_statistics"),
+					resource.TestCheckNoResourceAttr(resourceName, "s3_settings.0.timestamp_column_name"),
```
Comment (lyle-nel) [inline aws/resource_aws_dms_endpoint_test.go]: I don't know how to get this test to pass, even though the response object clearly indicates that the attribute is not set.

**Comment by lyle-nel** [comment]
I am using `resource_aws_elastic_beanstalk_application_test.go` as a reference, but I can't get this test to pass even though the response object clearly shows these attributes are not set. I must be missing something obvious, so I need someone more experienced in this to guide me in the right direction.
https://github.com/lyle-nel/terraform-provider-aws/blob/8a84842b721b054e74af6648f62d0b84c3bd2d59/aws/resource_aws_dms_endpoint_test.go#L77

**Comment by gerovermaas** [comment]
I have tested the version and can confirm that it now provision correct and I'm able to generate parquet files. Used the following resource definition:
```
resource "aws_dms_endpoint" "s3_target" {
  endpoint_id   = "s3-target-endpoint"
  endpoint_type = "target"
  engine_name   = "s3"

  extra_connection_attributes = "includeOpForFullLoad=true"

  s3_settings {
    service_access_role_arn = aws_iam_role.s3_target.arn
    bucket_folder           = "somefolder"
    bucket_name             = aws_s3_bucket.target.id
    data_format             = "parquet"
    timestamp_column_name   = "updateTS"
  }
}
```

And `aws dms describe-endpoints` shows:
```
        {
            "EndpointIdentifier": "s3-target-endpoint",
            "EndpointType": "TARGET",
            "EngineName": "s3",
            "EngineDisplayName": "Amazon S3",
            "ExtraConnectionAttributes": "bucketFolder=somefolder;bucketName=dms-poc-target2;compressionType=NONE;csvDelimiter=,;csvRowDelimiter=\\n;",
            "Status": "active",
            "EndpointArn": "arn:aws:dms:eu-west-2:627871771344:endpoint:ZCD2MXHTZVRTJUH2YSNWVGQ4JA",
            "SslMode": "none",
            "ServiceAccessRoleArn": "arn:aws:iam::627871771344:role/dms_test_s3_target_role",
            "S3Settings": {
                "ServiceAccessRoleArn": "arn:aws:iam::627871771344:role/dms_test_s3_target_role",
                "CsvRowDelimiter": "\\n",
                "CsvDelimiter": ",",
                "BucketFolder": "somefolder",
                "BucketName": "dms-poc-target2",
                "CompressionType": "NONE",
                "DataFormat": "parquet",
                "EnableStatistics": true
            }
        },
```

Would be great if this PR can be merged, it would help us a lot!

**Comment by hagridaaron** [comment]
This would also help me out quite a bit. Thank you so much @lyle-nel for all the hard work!

**Comment by gerovermaas** [comment]
We have been doing some more testing and although the `dataFormat=parquet` argument is now properly processed, we noticed that `timestamp_column_name` in the `s3_settings` is still not effectuated.  We used this resource definition:

```
resource "aws_dms_endpoint" "s3_target" {
  endpoint_id   = "s3-target-endpoint"
  endpoint_type = "target"
  engine_name   = "s3"

  s3_settings {
    service_access_role_arn = aws_iam_role.s3_target.arn
    bucket_folder           = "dms"
    bucket_name             = aws_s3_bucket.target.id
    data_format             = "parquet"
    timestamp_column_name   = "updateTS"
    compression_type = "GZIP"
  }
}
```

And `aws dms describe-endpoints` shows:
```
       {
            "EndpointIdentifier": "s3-target-endpoint",
            "EndpointType": "TARGET",
            "EngineName": "s3",
            "EngineDisplayName": "Amazon S3",
            "ExtraConnectionAttributes": "bucketFolder=dms;bucketName=dms-poc-target2;compressionType=GZIP;csvDelimiter=,;csvRowDelimiter=\\n;",
            "Status": "active",
            "EndpointArn": "arn:aws:dms:eu-west-2:xxxxxxxxx:endpoint:LCRRXDBE5PAJNUBGNT35ITUVNU",
            "SslMode": "none",
            "ServiceAccessRoleArn": "arn:aws:iam::xxxxxxxxx:role/dms_test_s3_target_role",
            "S3Settings": {
                "ServiceAccessRoleArn": "arn:aws:iam::xxxxxxxx:role/dms_test_s3_target_role",
                "CsvRowDelimiter": "\\n",
                "CsvDelimiter": ",",
                "BucketFolder": "dms",
                "BucketName": "dms-poc-target2",
                "CompressionType": "GZIP",
                "DataFormat": "parquet",
                "EnableStatistics": true
            }
        },
```

When we do an additional `terraform apply` with `  extra_connection_attributes = "TimestampColumnName=updateTSs"` included in the resource, then the describe shows:
```
        {
            "EndpointIdentifier": "s3-target-endpoint",
            "EndpointType": "TARGET",
            "EngineName": "s3",
            "EngineDisplayName": "Amazon S3",
            "ExtraConnectionAttributes": "compressionType=none;csvDelimiter=,;csvRowDelimiter=\\n;maxFileSize=1024000000;TimestampColumnName=updateTSs;",
            "Status": "active",
            "EndpointArn": "arn:aws:dms:eu-west-2:627871771344:endpoint:LCRRXDBE5PAJNUBGNT35ITUVNU",
            "SslMode": "none",
            "ServiceAccessRoleArn": "arn:aws:iam::627871771344:role/dms_test_s3_target_role",
            "S3Settings": {
                "ServiceAccessRoleArn": "arn:aws:iam::627871771344:role/dms_test_s3_target_role",
                "CsvRowDelimiter": "\\n",
                "CsvDelimiter": ",",
                "BucketFolder": "",
                "BucketName": "",
                "CompressionType": "none",
                "DataFormat": "parquet",
                "EnableStatistics": true
            }
        },
```

Is this something you can look into @lyle-nel ?

**Comment by lyle-nel** [comment]
@gerovermaas Yes, I know what the problem is.

My sincere apologies for this folks, once I understand what I am doing wrong with this one test, the test would expose issues such as this. I am hoping to figure out how this part needs to work soon, but some guidance from someone with experience in developing providers would go a long way. In the meanwhile, I have pushed the fix.

@gerovermaas , thanks for your valuable feedback and continued patience.

**Comment by gerovermaas** [comment]
Thank @lyle-nel ! I tested the new version and now the `timestamp_column_name` is properly handled. Great!

Used this definition and I'm happy we do not need the `extra_connection_attributes` anymore:
```
resource "aws_dms_endpoint" "s3_target" {
  endpoint_id   = "s3-target-endpoint"
  endpoint_type = "target"
  engine_name   = "s3"

  s3_settings {
    service_access_role_arn = aws_iam_role.s3_target.arn
    bucket_folder           = "dms"
    bucket_name             = aws_s3_bucket.target.id
    data_format             = "parquet"
    timestamp_column_name   = "updateTS"
    compression_type        = "GZIP"
  }
}```

**Comment by lyle-nel** [comment]
@bflad I see that you are one of the more recent contributors to this component. I was hoping you could have a quick look at this? Am I missing something very obvious that is causing the terraform state to contain the optional attributes even though they are not being set?

**Comment by lyle-nel** [comment]
I am closing this pull request until I can figure out how to fix the test.

**Pull request closed without merging**